service/s3/statusok_error.go

```diff
@@ -21,6 +21,8 @@ func copyMultipartStatusOKUnmarhsalError(r *request.Request) {
 		)
 		return
 	}
+	r.HTTPResponse.Body.Close()
+
```

Comment:
I tried it, but, as I explained in the PR description, looks like we don't actually want to close it in case of error because later Unmarshal handlers expect it not to be closed and close it themselves. One of the tests fails in this case, so I opted to keep the current behavior of not closing it. Let me know if you think it should be changed, although that would require adding more logic to the Unmarshal handlers.

Reply:
Specifically, this is the test case failure that happens if we close response after io.ReadAll returns error:
```
--- FAIL: TestStatusOKPayloadHandling (0.01s)
    --- FAIL: TestStatusOKPayloadHandling/200_error_partial_response (0.00s)
        statusok_error_test.go:383: expect unexpected EOF error in, SerializationError: failed to decode REST XML response
            	status code: 200, request id: , host id: 
            caused by: http: read on closed response body
```

Reply:
Which comes from this Unmarshal handler: https://github.com/aws/aws-sdk-go/blob/v1.40.15/private/protocol/restxml/restxml.go#L56

Reply:
(note, it correctly closes the body irrespective of the error)

Reply:
Thanks for the update. The `Umarshaler` handling this error probably is unexpected behavior, and an artifact of the SDK's request middleware always going to the next handler in the chain regardless of error. The `restxml.Unmarshaler` wouldn't be able to do anything with the failed `r.HTTPResonse.Body`, and is most likely ignoring or stomping on top of the error returned by `ioutil.ReadAll`.

Since the `r.HTTPResponse.Body` will be in an unusable state, if `ReadAll` fails, we probably want to replace the `r.HTTPResponse.Body` with a empty no-op closer to prevent downstream consumers from tripping on the invalid body.  I think that should address both the unexpected error in restxml.Unmarshal and closing the body in case of ReadAll error.

```go
func copyMultipartStatusOKUnmarshalError(r *request.Request) {
	defer r.HTTPResponse.Body.Close()
	b, err := ioutil.ReadAll(r.HTTPResponse.Body)
	if err != nil {
		r.Error = awserr.NewRequestFailure(
			awserr.New(request.ErrCodeSerialization, "unable to read response body", err),
			r.HTTPResponse.StatusCode,
			r.RequestID,
		)
		r.HTTPResponse.Body = ioutil.NopCloser(bytes.NewBuffer(nil))
		return
	}
	
	//...
```

We should also consider updating `restxml.Unmarshal` to not attempt to deserialize the body if there is already a `r.Error` value on the `request.Request`. This is a bigger change/risk since `restxml.Unmarshal` is used by several services. We'd need to do a deep dive investigating that change more.

Reply:
Yes ok that makes sense. I'll update the diff. 

One question - in this case `defer` muddies the logic a bit as we want to always close right after ReadAll. Unless you think that this is worth keeping as a recognizable pattern, I'd get rid of it and just move the closing explicitly after the ReadAll. What do you think?

Reply:
Thanks for updating the close. Moving the close directly after `ReadAll` looks good, and removes need for defer.

Reply:
Thank you!